### PR TITLE
AzureBlob: Download the content in chunks

### PR DESCRIPTION
**Pull request by clguimanMSFT**
* Add a JsonParser class that can parse a JSON stream
* Add BlobReader class that reads the blob in chunks
* Refactor & simplify the process method
* Add tests for the azure blob input plugin

 #70

### CODE REVIEW

File: Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb [hunk 1]
```diff
@@ -142,126 +151,100 @@ def stop
...
+            })
           else
-            @logger.debug("Non-json codec or the policy is do not break")
-            # Putting header and content and tail together before pushing into event queue
-            content = "#{header}#{content}" unless header.nil? || header.length == 0
-            @codec.decode(content) do |event|
-              decorate(event)
-              queue << event
-            end # decode
+            begin
```
Comment (brahmnes) [inline Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb]: If the codec is not json, you still need to respect that codec. It looks like your change in this else clause is not using the codec at all.

File: Logstash/logstash-input-azureblob/lib/com/microsoft/json-parser.rb [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+  
+        if depth == 0
+          stream_end_offset = @parser.getLocation() .getStreamOffset() - 1
+          @logger.debug ("JsonParser::parse_single_object Json  object found stream_start_offset: #{stream_start_offset} stream_end_offset: #{stream_end_offset}")
+  
+          on_json_cbk.call(@stream_reader.get_stream_buffer(stream_start_offset - @stream_base_offset,  stream_end_offset - @stream_base_offset))
+          stream_start_offset = stream_end_offset + 1
+  
+          #Drop parsed bytes
+          @stream_reader.drop_stream(stream_end_offset  - @stream_base_offset)
```
Comment (brahmnes) [inline Logstash/logstash-input-azureblob/lib/com/microsoft/json-parser.rb]: I assume this call is pretty fast and we do this to trim down on working memory usage?

File: Logstash/logstash-input-azureblob/lib/com/microsoft/json-parser.rb [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        depth = depth + 1
+      elsif event == javax::json::stream::JsonParser::Event::END_OBJECT
+        depth = depth - 1 # can't be negative because the parser handles the format correctness
+  
+        if depth == 0
+          stream_end_offset = @parser.getLocation() .getStreamOffset() - 1
+          @logger.debug ("JsonParser::parse_single_object Json  object found stream_start_offset: #{stream_start_offset} stream_end_offset: #{stream_end_offset}")
+  
+          on_json_cbk.call(@stream_reader.get_stream_buffer(stream_start_offset - @stream_base_offset,  stream_end_offset - @stream_base_offset))
+          stream_start_offset = stream_end_offset + 1
```
Comment (brahmnes) [inline Logstash/logstash-input-azureblob/lib/com/microsoft/json-parser.rb]: Looks like you can eliminate one of stream_start_offset and stream_offset. For example, delete line 75 and rename line 62 to be stream_start_offset

File: Logstash/logstash-input-azureblob/logstash-input-azureblob.gemspec [hunk 4]
```diff
@@ -23,4 +24,10 @@ Gem::Specification.new do |s|
   s.add_runtime_dependency 'stud', '>= 0.0.22'
   s.add_runtime_dependency 'azure-storage', '~> 0.12.3.preview'
   s.add_development_dependency 'logstash-devutils'
+  s.add_development_dependency 'logging'
+
+  # Jar dependencies
+  #s.requirements << "jar 'javax.json:javax.json-api', '1.1'"
+  s.requirements << "jar 'org.glassfish:javax.json', '1.1'"
```
Comment (brahmnes) [inline Logstash/logstash-input-azureblob/logstash-input-azureblob.gemspec]: What's the new thing that this other parser can and the previous cannot?

Comment (clguimanMSFT) [reply Logstash/logstash-input-azureblob/logstash-input-azureblob.gemspec]: this is just an artifact. The commented line was just my first try in getting the JsonParser dependency (it was the wrong namespace). Previously we didn't have any parser, just a simple string check

File: Logstash/logstash-input-azureblob/lib/com/microsoft/json-parser.rb (hunk 3 above)
Comment (clguimanMSFT) [reply Logstash/logstash-input-azureblob/lib/com/microsoft/json-parser.rb]: you're right. That was left there by accident when I had some debug lines

File: Logstash/logstash-input-azureblob/lib/com/microsoft/json-parser.rb (hunk 2 above)
Comment (clguimanMSFT) [reply Logstash/logstash-input-azureblob/lib/com/microsoft/json-parser.rb]: yes

**Pull request merged**